## Занятие 5. Линейные модели для классификации

Веберите любые данные для задачи классификации из репозитория данных для машинного обучения (UCI Machine learning repository: http://archive.ics.uci.edu/ml/index.php) или возьмите свои данные и постройте линейные модели в соответствии со следующей схемой. Комментарии к каждому разделу обязательны.

*Выполнил:* Бронников Егор ПМ-1901

*Последнее изменение:* 13/12/2021

### Импорты

In [1]:
from pandas import read_csv
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.svm import SVC

### Данные

*Источник:* https://github.com/jbrownlee/Datasets

*Название:* **Cancer (Haberman's)**

### Описание данных

Данные о пациентах больных раком.

### Описание полей

1. **age** — возраст пациента;
1. **operation_year** — год операции 19##;
1. **axillary_nodes** — количество обнаруженных положительных узлов;
1. **survival_status** — выживаемость после операции:
    - 1 $-$ пациент прожил 5 лет и более 
    - 2 $-$ пациент умер в течение 5 лет

Смотрим на данные.

In [2]:
filename = "haberman.csv"
names = ["age", "operation_year", "axillary_nodes", "survival_status"]
dataframe = read_csv(filename, names=names)
array = dataframe.values
dataframe.head()

age  operation_year  axillary_nodes  survival_status
0   30              64               1                1
1   30              62               3                1
2   30              65               0                1
3   31              59               2                1
4   31              65               4                1

In [3]:
dataframe.describe()

age  operation_year  axillary_nodes  survival_status
count  306.000000      306.000000      306.000000       306.000000
mean    52.457516       62.852941        4.026144         1.264706
std     10.803452        3.249405        7.189654         0.441899
min     30.000000       58.000000        0.000000         1.000000
25%     44.000000       60.000000        0.000000         1.000000
50%     52.000000       63.000000        1.000000         1.000000
75%     60.750000       65.750000        4.000000         2.000000
max     83.000000       69.000000       52.000000         2.000000

In [4]:
X = array[:,0:3]
Y = array[:,3]

### Logistic Regression

In [16]:
kfold = KFold(n_splits=10, random_state=7, shuffle=True)
model = LogisticRegression(solver='liblinear')
results = cross_val_score(model, X, Y, cv=kfold)
print(f"{round(results.mean()*100, 6)} %")

74.83871 %


Значение точности вычислений составляет **75%**, это означает, что модель логистической регрессии хорошо подходит для классификации.

### Linear Discriminant Analysis

In [17]:
kfold = KFold(n_splits=10, random_state=7, shuffle=True)
model = LinearDiscriminantAnalysis()
results = cross_val_score(model, X, Y, cv=kfold)
print(f"{round(results.mean()*100, 6)} %")

74.83871 %


Линейный дискриминантный анализ показал точность классификации **75%**, что совпало с моделью логистической регрессии.

### Support Vector Machines

In [18]:
kfold = KFold(n_splits=10, random_state=7, shuffle=True)
model = SVC()
results = cross_val_score(model, X, Y, cv=kfold)
print(f"{round(results.mean()*100, 6)} %")

72.870968 %


Линейный метод опорных векторов показал точность классификации **73%**, что чуть хуже предыдущих результатов, но всё раво не так плохо.

### Выводы

Таким образом, лучше всего подходят модель логистической регрессии и линейный дискриминантный анализ для классификации на данной выборке данных.